### PR TITLE
Improving stale container cleanup script

### PR DESCRIPTION
**Pull request by pabloem**
This is the expected outcome from running this script today: https://gist.github.com/pabloem/95cff6aca3e17fd09f4c792a54d14dde

Should we run it?

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by pabloem** [comment]
r: @Abacn

**Comment by Abacn** [comment]
Thanks for the quick solution!

Two concerns:
the 5 day filter is fine for prebuilt_sdk but may be aggressive for beam-sdk and beam_portability, given that the docker publish jobs have history to have failed for a couple of days time range: https://ci-beam.apache.org/job/beam_Publish_Beam_SDK_Snapshots/4095/  I would suggest a 10 day or 30 day filter for beam-sdk and beam_portability;

also we may want to keep some latest but no longer produced images as our nightly napshot doing: https://repository.apache.org/content/groups/snapshots/org/apache/beam/ for example, flink-1.12 is no longer generated but we may still want to preserve the latest 5 (or other number) days containers.

**Comment by pabloem** [comment]
that makes sense to me. I'll look into it...

**Comment by pabloem** [comment]
The difference in the expected outcome is that 6 fewer images are deleted:

```
diff new_expected_outcome expected_outcome  | grep Delete
> Deleting image. Command: gcloud container images delete gcr.io/apache-beam-testing/beam-sdk/beam_python3.6_sdk@sha256:3f0ff833838b5945f1a9551a51f6d3223d7f11d4777360c678641c12c1995862 --force-delete-tags -q
> Deleting image. Command: gcloud container images delete gcr.io/apache-beam-testing/beam_portability/beam_flink1.12_job_server@sha256:911283585036550af2b3ab86dd87aab88a229103f63aabf080ebec9be3589a3b --force-delete-tags -q
> Deleting image. Command: gcloud container images delete gcr.io/apache-beam-testing/beam_portability/beam_go_sdk@sha256:de12cb02d386c0efa1799b39c9a8a9f5035ad1bf29248440909e2223fea8c157 --force-delete-tags -q
> Deleting image. Command: gcloud container images delete gcr.io/apache-beam-testing/beam_portability/beam_java_sdk@sha256:a602d9c25750a6c1351f7703977e03d5a45a0e7bb762958473ef68a873d1eacd --force-delete-tags -q
> Deleting image. Command: gcloud container images delete gcr.io/apache-beam-testing/beam_portability/beam_python2.7_sdk@sha256:4feefecb1b41a4c7273e46266fab06a6215e6b985613bab88921a6edbcbdd48e --force-delete-tags -q
> Deleting image. Command: gcloud container images delete gcr.io/apache-beam-testing/beam_portability/beam_python3.6_sdk@sha256:106ece6c7d8bd11f35250b72c71674735065b176332ecfffc01aa86a47504c7b --force-delete-tags -q
```

**Comment by Abacn** [comment]
Thanks again. Looking into the gist file I see this log:
```
IMAGES FOR image gcr.io/apache-beam-testing/beam-sdk/beam_python3.9_sdkgcr.io/apache-beam-testing/beam_portability/beam_flink1.10_job_server
Command gcloud container images list-tags gcr.io/apache-beam-testing/beam-sdk/beam_python3.9_sdkgcr.io/apache-beam-testing/beam_portability/beam_flink1.10_job_server --sort-by=TIMESTAMP --filter=timestamp.datetime < 2022-11-03T14:18:54-07:00 --format=get(digest)
IMAGES FOR image gcr.io/apache-beam-testing/beam_portability/beam_flink1.12_job_server
```
looks like for some reason image name are concatenated and Python3.9 images are not listed

**Comment by pabloem** [comment]
wow great catch : )

**Comment by pabloem** [comment]
I've tested this locally and your fix works as expeted (now all the image categories are parsed properly)

**Comment by pabloem** [comment]
if this LGTY, I will run it on my machine (since the first run will delete many many containers) - and merge it

File: .test-infra/tools/stale_dataflow_prebuilt_image_cleaner.sh [hunk 1]
```diff
@@ -19,15 +19,38 @@
...
+  IMAGE_NAMES+=" "
+done
+
+echo $IMAGE_NAMES
+
+for image_name in ${IMAGE_NAMES[@]}; do
+  echo IMAGES FOR image ${image_name}
+  echo "Command" gcloud container images list-tags \
+  ${image_name} \
+  --sort-by=TIMESTAMP  --filter="timestamp.datetime < $(date --iso-8601=s -d '5 days ago')" \
```
Comment (Abacn) [inline .test-infra/tools/stale_dataflow_prebuilt_image_cleaner.sh:38]:
```suggestion
  --sort-by=TIMESTAMP  --filter="NOT tags:latest AND timestamp.datetime < $(date --iso-8601=s -d '5 days ago')" \
```

Comment (Abacn) [reply .test-infra/tools/stale_dataflow_prebuilt_image_cleaner.sh:38]: nit (does not affect functionality)

**Review by Abacn — COMMENTED** [review]
LGTM

**Comment by pabloem** [comment]
hm I've just realized that this script is super slow if there's a lot of containers to delete (maybe it's okay?)

**Comment by Abacn** [comment]
btw if there is a worry we can reduce the frequency of this job to daily. Currently it runs 6 times a day which is a lot.

**Comment by pabloem** [comment]
thanks Yi! Let's see how it behaves over the next few days

**Pull request merged**